File: core/src/main/java/nl/sjtek/control/core/modules/NightMode.java
package nl.sjtek.control.core.modules;

import nl.sjtek.control.core.network.ApiHandler;
import nl.sjtek.control.core.network.Arguments;
import nl.sjtek.control.core.utils.Speech;
import nl.sjtek.control.data.responses.NightModeResponse;
import nl.sjtek.control.data.responses.Response;

import java.util.Calendar;
import java.util.Date;
import java.util.Timer;
import java.util.TimerTask;

/**
 * Created by wouter on 28-11-15.
 */
public class NightMode extends BaseModule {

    private boolean enabled = false;
    private Timer timer;
    private Date nextDisable = Calendar.getInstance().getTime();

    public NightMode(String key) {
        super(key);
        this.timer = new Timer();
        schedule();
    }

    public void enable(Arguments arguments) {
        ApiHandler.getInstance().getMusic().pause(new Arguments());
        if (arguments.useVoice()) Speech.speak("Entering night mode");
        enabled = true;
        dataChanged();
    }

    public void disable(Arguments arguments) {
        enabled = false;
        dataChanged();
        if (arguments.useVoice()) Speech.speak("Disabling night mode. Good morning");
    }

    @Override
    public Response getResponse() {
        return new NightModeResponse(enabled, nextDisable.toString());
    }

    @Override
    public String getSummaryText() {
        return "Night mode is " + (enabled ? "enabled" : "disabled") + ".";
    }

    @Override
    public boolean isEnabled(String user) {
        return false;
    }

    private void schedule() {
        Calendar calendar = Calendar.getInstance();
        calendar.set(Calendar.HOUR_OF_DAY, 7);
        calendar.set(Calendar.MINUTE, 0);
        calendar.set(Calendar.SECOND, 0);
        calendar.add(Calendar.DAY_OF_YEAR, 1);
        nextDisable = calendar.getTime();
        timer.schedule(new DisableTask(), nextDisable);
    }

    private class DisableTask extends TimerTask {

        @Override
        public void run() {
            enabled = false;
            schedule();
        }
    }
}

File: core/src/main/kotlin/nl/sjtek/control/core/modules/NightMode.kt
package nl.sjtek.control.core.modules

import nl.sjtek.control.core.events.Bus
import nl.sjtek.control.core.events.NightModeEvent
import nl.sjtek.control.core.response.ResponseCache
import nl.sjtek.control.data.response.NightMode
import nl.sjtek.control.data.response.Response
import spark.Request
import spark.Spark.get
import spark.Spark.path
import spark.Response as SparkResponse

class NightMode(key: String) : Module(key) {

    override val response: Response
        get() = NightMode(key, enabled)
    private var enabled: Boolean = false
        set(value) {
            field = value
            Bus.post(NightModeEvent(value))
            ResponseCache.post(this)
        }

    override fun initSpark() {
        path("/nightmode") {
            get("/enable", this::enable)
            get("/disable", this::disable)
            get("/toggle", this::toggle)
        }
    }

    private fun enable(req: Request, res: SparkResponse) {
        enabled = true
    }

    private fun disable(req: Request, res: SparkResponse) {
        enabled = false
    }

    private fun toggle(req: Request, res: SparkResponse) {
        enabled = !enabled
    }
}